src/k_smallest.rs

```diff
@@ -1,20 +1,214 @@
+{
+    k_smallest_general(iter, k, cmp)
+}
+
+pub(crate) fn k_smallest_by_key<T, I, F, K>(iter: I, k: usize, key: F) -> IntoIter<T>
+where
+    I: Iterator<Item = T>,
+    F: Fn(&T) -> K,
+    K: Ord,
+{
+    let iter = iter.map(|v| Pair(key(&v), v));
+
+    let results: Vec<_> = k_smallest_general(iter, k, Ord::cmp)
+        .map(|Pair(_, t)| t)
+        .collect();
+    results.into_iter()
+}
+
+/// An efficient heapsort requires that the heap ordering is the inverse of the desired sort order
+/// So the basic case of retrieving minimum elements requires a max heap
+///
+/// This type does not attempt to reproduce all the functionality of [std::collections::BinaryHeap] and instead only implements what is needed for iter operations,
+/// e.g. we do not need to insert single elements.
+/// Additionally, some minor optimizations used in the std BinaryHeap are not used here, e.g. elements are actually swapped rather than managing a "hole"
+struct MaxHeap<T, C> {
+    // It may be not be possible to shrink the storage for smaller sequencess
+    // so manually manage the initialization
+    // This is **assumed not to be empty**
+    storage: alloc::vec::Vec<T>,
+    comparator: C,
+    // this is always less or equal to the count of actual elements
+    // allowing it to be less means the heap property can cover only a subset of the vec
+    // while reusing the storage
+    len: usize,
+}
+
+impl<T, C> MaxHeap<T, C>
+where
+    C: Fn(&T, &T) -> Ordering,
+{
+    fn from_iter<I>(k: usize, comparator: C, mut iter: I) -> Self
+    where
+        I: Iterator<Item = T>,
+    {
+        let storage: Vec<T> = iter.by_ref().take(k).collect();
+
+        let mut heap = Self {
+            len: storage.len(),
+            storage,
+            comparator,
+        };
+        // Filling up the storage and only afterwards rearranging to form a valid heap is slightly more efficient
+        // (But only by a factor of lg(k) and I'd love to hear of a usecase where that matters)
+        heap.heapify();
+
+        if k == heap.storage.len() {
+            // Nothing else needs done if we didn't fill the storage in the first place
+            // Also avoids unexpected behaviour with restartable iterators
+            for val in iter {
+                heap.push_pop(val);
+            }
+        }
+        heap
+    }
+
+    /// Retrieves the element at the given index.
+    /// Returning an Option means that bounds checking is subsumed in checking the
+    /// element ordering.
+    fn get(&self, index: usize) -> Option<&T> {
+        if index < self.len {
+            self.storage.get(index)
+        } else {
+            None
+        }
+    }
 
-    let mut heap = iter.by_ref().take(k).collect::<BinaryHeap<_>>();
+    /// Sift the element currently at `origin` **away** from the root until it is properly ordered
+    fn sift_down(&mut self, origin: usize) {
+        fn children_of(n: usize) -> (usize, usize) {
+            (2 * n, 2 * n + 1)
```

Comment:
Is this really correct? This looks dubious, as index `0` is mapped to `(0,1)`, while I think it should be `(1,2)`.

Reply:
I'd agree. It worked, but I think the result was a slightly misshapen tree. Has now been corrected